File: clients/java/client/src/main/java/ai/sunra/client/SunraClient.java
package ai.sunra.client;

import ai.sunra.client.queue.QueueClient;
import jakarta.annotation.Nonnull;
import jakarta.annotation.Nullable;

/**
 * The main client interface for interacting with the sunra APIs.
 *
 * @see #withConfig(ClientConfig) method to create a new client instance
 * with the provided configuration.
 * @see #withEnvCredentials() method to create a new client instance
 * with the credentials resolved from the environment.
 */
public interface SunraClient {
    /**
     * Subscribe to the specified endpoint with the provided options. This method is recommended for
     * long-running operations. The subscription will return the result once the operation is
     * completed.
     *
     * @param <O> Output type.
     * @param endpointId The endpoint ID to subscribe to, e.g. `sunra/lcm`.
     * @param options The subscribe options.
     * @return The result of the operation, or null if onError callback is provided and an error occurs.
     * @see #queue()
     */
    @Nullable
    <O> Output<O> subscribe(String endpointId, SubscribeOptions<O> options);

    /**
     * Get the queue client for interacting with the sunra queue.
     *
     * @return The queue client.
     */
    QueueClient queue();

    /**
     * Create a new client instance with the provided configuration.
     *
     * @param config The client configuration.
     * @return The new client instance.
     */
    static SunraClient withConfig(@Nonnull ClientConfig config) {
        return new SunraClientImpl(config);
    }

    /**
     * Create a new client instance with the credentials resolved from the `SUNRA_KEY` environment
     * variable.
     *
     * @return The new client instance.
     */
    static SunraClient withEnvCredentials() {
        return new SunraClientImpl(ClientConfig.withCredentials(CredentialsResolver.fromEnv()));
    }

    /**
     * Create a new client instance with the provided proxy URL. With this configuration all
     * requests will be proxied through the provided URL and the sunra target url will be in a request
     * header called `X-Sunra-Target-Url`.
     *
     * @param proxyUrl The proxy URL.
     * @return The new client instance.
     */
    static SunraClient withProxyUrl(@Nonnull String proxyUrl) {
        return new SunraClientImpl(ClientConfig.builder().withProxyUrl(proxyUrl).build());
    }
}

File: clients/java/client-kotlin/bin/main/SunraClient.kt
package ai.sunra.client.kt

import ai.sunra.client.AsyncSunraClient
import ai.sunra.client.ClientConfig
import ai.sunra.client.CredentialsResolver
import ai.sunra.client.queue.QueueStatus
import com.google.gson.JsonObject
import kotlinx.coroutines.future.await
import kotlin.reflect.KClass
import ai.sunra.client.RunOptions as InternalRunOptions
import ai.sunra.client.SubscribeOptions as InternalSubscribeOptions

data class RunOptions(
    val httpMethod: String = "POST",
)

data class SubscribeOptions(
    val logs: Boolean = false,
    val webhookUrl: String? = null,
)

/**
 * The main client class that provides access to simple API model usage,
 * as well as access to the [queue] APIs.
 * @see AsyncSunraClient
 */
interface SunraClient {
    /** The queue client with specific methods to interact with the queue API.
     *
     * **Note:** that the [subscribe] method is a convenience method that uses the
     * [queue] client to submit a request and poll for the result.
     */
    val queue: QueueClient

    /**
     * Submits a request to the given [endpointId]. This method
     * uses the [queue] API to submit the request and poll for the result.
     *
     * This is useful for long running requests, and it's the preffered way
     * to interact with the model APIs.
     *
     * @param endpointId The ID of the endpoint to send the request to.
     * @param input The input data to send to the endpoint.
     * @param resultType The expected result type of the request.
     * @param options The options to use for the request.
     * @param onQueueUpdate A callback to receive status updates from the queue subscription.
     */
    suspend fun <Output : Any> subscribe(
        endpointId: String,
        input: Any,
        resultType: KClass<Output>,
        options: SubscribeOptions = SubscribeOptions(),
        onQueueUpdate: OnStatusUpdate? = null,
    ): RequestOutput<Output>
}

/**
 * A callback for receiving status updates from a queue subscription.
 */
typealias OnStatusUpdate = (update: QueueStatus.StatusUpdate) -> Unit

/**
 * A Kotlin implementation of [SunraClient] that wraps the Java [AsyncSunraClient].
 */
internal class SunraClientKotlinImpl(
    config: ClientConfig,
) : SunraClient {
    private val client = AsyncSunraClient.withConfig(config)

    override val queue: QueueClient = QueueClientImpl(client.queue())

    override suspend fun <Output : Any> subscribe(
        endpointId: String,
        input: Any,
        resultType: KClass<Output>,
        options: SubscribeOptions,
        onQueueUpdate: OnStatusUpdate?,
    ): RequestOutput<Output> {
        println(resultType)
        println(options)
        val internalOptions =
            InternalSubscribeOptions.builder<Output>()
                .input(input)
                .resultType(resultType.java)
                .logs(options.logs)
                .onQueueUpdate(onQueueUpdate)
                .build()
        return client.subscribe(endpointId, internalOptions).thenConvertOutput().await()
    }
}

suspend inline fun <reified Output : Any> SunraClient.subscribe(
    endpointId: String,
    input: Any,
    options: SubscribeOptions = SubscribeOptions(),
    noinline onUpdate: OnStatusUpdate? = null,
) = this.subscribe(endpointId, input, Output::class, options, onUpdate)

@JvmName("subscribe_")
suspend inline fun SunraClient.subscribe(
    endpointId: String,
    input: Any,
    options: SubscribeOptions = SubscribeOptions(),
    noinline onUpdate: OnStatusUpdate? = null,
) = this.subscribe(endpointId, input, JsonObject::class, options, onUpdate)

fun createSunraClient(config: ClientConfig? = null): SunraClient =
    SunraClientKotlinImpl(config ?: ClientConfig.withCredentials(CredentialsResolver.fromEnv()))

fun createSunraClient(credentialsResolver: CredentialsResolver): SunraClient =
    SunraClientKotlinImpl(ClientConfig.withCredentials(credentialsResolver))
